src/librustc/middle/const_eval.rs

```diff
@@ -1131,19 +891,22 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
         if !tcx.sess.features.borrow().const_indexing {
             signal!(e, IndexOpFeatureGated);
         }
-        let arr_hint = ty_hint.erase_hint();
-        let arr = try!(eval_const_expr_partial(tcx, arr, arr_hint, fn_args));
+        // FIXME: ty_hint is not really correct here, it should be [ty_hint; $something]
+        let arr = try!(eval_const_expr_partial(tcx, arr, ty_hint, fn_args));
         let idx_hint = ty_hint.checked_or(tcx.types.usize);
         let idx = match try!(eval_const_expr_partial(tcx, idx, idx_hint, fn_args)) {
-            Int(i) if i >= 0 => i as u64,
-            Int(_) => signal!(idx, IndexNegative),
-            Uint(i) => i,
+            Integral(Usize(i)) => i.as_u64(tcx.sess.target.uint_type),
+            Integral(Infer(i)) => i as u64,
+            Integral(_) => signal!(idx, IndexNegative),
```

Comment:
Put another way: we don't know that the index is actually negative, right? Is this code even reachable? If so, when? Wouldn't "bad type" be more accurate? (If this code only runs for errors, maybe we should just try to cast it to usize and go on with our lives, presuming an error is already reported.)